### ISSUE TITLE
False positive `unnecessary-dunder-call` for `super().__init__(arg)`

### ISSUE THREAD

**Issue by schrodervictor** [PyCQA/pylint#6889]
### Bug description

Consider a class with some `__init__` dunder method:
```
class Foo:

    def __init__(self, data):
        self.data = data
```

And a child class extending from it, relying on the parent `__init__` method:
```
class Product(Foo):

    def __init__(self, data):
        super().__init__(data)
        self.whatever = 'something'
```

Calling `super().__init__(data)` is Python's idiomatic way to bring the parent initialization to the instance of the child class being instantiated and keep the hierarchy chain correctly set. It isn't an unnecessary call to a dunder method. However, the current version of pylint (`2.14.1`) gives `unnecessary-dunder-call` for the code above, with the message `"Unnecessarily calls dunder method __init__. Instantiate class directly."`. It was not the case in our previous build, using pylint version `2.13.9`.

### Command used

```shell
pylint --rcfile=app/.pylintrc app
```

### Pylint output

```shell
pylint --rcfile=app/.pylintrc app
************* Module our_module.product
/app/our_module/product.py:62:8: C2801: Unnecessarily calls dunder method __init__. Instantiate class directly. (unnecessary-dunder-call)
```

### Expected behavior

Calling `super().__init__(data)` is the expected way to inherit and extend the initialization method from the parent class. It should pass cleanly.

### Pylint version

```shell
pylint 2.14.1
astroid 2.11.5
Python 3.10.2 (main, Jan 29 2022, 03:40:37) [GCC 10.3.1 20211027]
```

### OS / Environment

OS: Alpine `v3.15` (from official docker image `python:3.10.2-alpine3.15`)
Terminal: BusyBox `v1.34.1` (2021-11-23 00:57:35 UTC) multi-call binary.

**Comment by Pierre-Sassoulas** [comment]
Thank you for opening the issue. I can't reproduce with the example you gave. Seeing that the error is line 62 in the output you provided I think we miss something from the original code that would make reproducing possible.

**Comment by schrodervictor** [comment]
Thanks for looking into it, @Pierre-Sassoulas. I'll provide a more comprehensive code example. Let me just trim it down to the minimal necessary until where I can still see the error.

**Comment by schrodervictor** [comment]
I took another look @Pierre-Sassoulas. It seems that the issue is completely different. In fact, there is a call to an `__init__` method under a non-dunder method. The call is done as `self.__init__(data)` and is necessary to rebind data into the currently existing instance without repeating initialization logic. There's no way to create a new instance directly in this case.

Curiously enough, this was not being reported by pylint `v2.13`. Should it really be the case? As far I understood, calls to dunder methods on `self` were exceptions for this rule, as they happen relatively often.

Code example:
```
class Product:

    # several methods, including __init__ and other dunder methods...

    def import_data(self, data: Dict[str, str]) -> None:
        temp_product: Product(data)
        new_data = merge_dicts(self.raw_view, temp_product.raw_view)
        self.__init__(new_data)
```

**Comment by Pierre-Sassoulas** [comment]
Why not change the instance attributes directly like this:
```python
    def import_data(self, data: Dict[str, str]) -> None:
        self.data = merge_dicts(self.raw_view, temp_product.raw_view)

old_product.import_data(data)
```       

Or create a new instance directly ?

```python
    def import_data(self, data: Dict[str, str]) -> "Product":
        temp_product: Product(data)
        new_data = merge_dicts(self.raw_view, temp_product.raw_view)
        return Product(new_data)
        
new_product = old_product.import_data(data)        
 ```

**Comment by Pierre-Sassoulas** [comment]
(This is a new check in 2.14)

**Comment by schrodervictor** [comment]
> Why not change the instance attributes directly like this:
> 
> ```python
>     def import_data(self, data: Dict[str, str]) -> None:
>         self.data = merge_dicts(self.raw_view, temp_product.raw_view)
> 
> old_product.import_data(data)
> ```
> 
> Or create a new instance directly ?
> 
> ```python
>     def import_data(self, data: Dict[str, str]) -> "Product":
>         temp_product: Product(data)
>         new_data = merge_dicts(self.raw_view, temp_product.raw_view)
>         return Product(new_data)
>         
> new_product = old_product.import_data(data)        
> ```

Thanks for the suggestions. Of course the code example above was simplified just to give an idea. The kind of mapping that happens on init and the affected attributes are much more complex than that. Also, because the instance reference is relevant and stored in instances of other objects, returning a new instance is not an option in this particular case. That part of the code wouldn't have access to all the places where a reference to the original instance is available. Trust me, having a fresh instance would have been my preferred approach too.

But we are just now discussing concrete implementation details. The fact that the check was recently introduced in v2.14 explains why we weren't seeing this before. I'll move forward and close this issue, as not related to pylint.

Thanks for the help and quick replies!
